### PR TITLE
Srk53/annihilate chaos

### PR DESCRIPTION
**Pull request by neongreen**

### CODE REVIEW

**Comment by neongreen** [comment]
I've made a PR out of this branch to make it easier to leave comments.

File: README.md [hunk 1]
```diff
@@ -4,105 +4,215 @@ Universum
 [![Build Status](https://travis-ci.org/serokell/universum.svg?branch=master)](https://travis-ci.org/serokell/universum)
 [![Hackage](https://img.shields.io/hackage/v/universum.svg)](https://hackage.haskell.org/package/universum)
 
-Prelude used in Serokell.
+Custom _Prelude_ used in Serokell.
```
Comment (neongreen) [inline README.md]: I think `Prelude` refers to a particular module, not to the concept, so writing “Custom prelude used in Serokell” would've looked more natural (at least for me).

File: README.md [hunk 2]
```diff
@@ -4,105 +4,215 @@ Universum
...
+1. Avoid all [partial functions](https://www.reddit.com/r/haskell/comments/5n51u3/why_are_partial_functions_as_in_head_tail_bad/).
+   We like [total](http://mathworld.wolfram.com/TotalFunction.html) and exception-free functions.
+   You can still use some [_unsafe_](https://github.com/serokell/universum/blob/master/src/Unsafe.hs)
+   functions but they are not exported by default.
+2. Use more efficient [string representations](https://www.reddit.com/r/haskell/comments/29jw0s/whats_wrong_with_string/).
+   `String` type is crushingly inefficient. All our functions tries to be polymorphic over string
+   types or use [`Text`](http://hackage.haskell.org/package/text-1.2.2.1/docs/Data-Text.html)
+   as the default string type. Though `String` type alias is still reexported because community
+   is evolving slowly, some libraries still use `String` type. But we recommend to avoid `String`!
+3. Don't reinvent wheels. We're not trying to rebuild whole type hierarchy from scratch
```
Comment (neongreen) [inline README.md]: Well, we are a bit, with `one` and `NontrivialContainer`.

Comment (neongreen) [inline README.md]: I mean, it's not reinventing wheels (strictly speaking), but it's similar to what `classy-prelude` does.

File: src/Print.hs [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
+
+module Print
        ( Print (..)
        , putText
        , putLText
        ) where
 
 import qualified Base
 import           Data.Function              ((.))
+import qualified Prelude                    as Prelude
```
Comment (neongreen) [inline src/Print.hs]: Just `import qualified Prelude`.

File: src/Universum.hs [hunk 4]
```diff
@@ -169,36 +174,76 @@ import           Lens.Micro               as X (Lens, Lens', Traversal, Traversa
...
 
+-- | Renamed version of 'Prelude.id'.
 identity :: a -> a
 identity x = x
 
+-- | 'Prelude.map' generalized to 'Functor'.
 map :: Functor f => (a -> b) -> f a -> f b
 map = fmap
 
+-- | Alias for @flip map@.
```
Comment (neongreen) [inline src/Universum.hs]: Perhaps let's remove `foreach`? It's not used anywhere in our code and it's unlikely that it ever will.

File: src/Universum.hs [hunk 5]
```diff
@@ -211,5 +256,14 @@ show x = X.fromString (PBase.show x)
 pretty :: Buildable a => a -> Text
 pretty = X.toStrict . prettyL
 
+-- | Similar to 'pretty' but for 'LText'.
 prettyL :: Buildable a => a -> LText
 prettyL = toLazyText . build
+
+-- | Lifted alias for 'Control.Exception.Base.evaluate' with clearer name.
+evaluateWHNF :: MonadIO m => a -> m a
```
Comment (neongreen) [inline src/Universum.hs]: Can you also add `evaluateNF_` and `evaluateWHNF_`?

```haskell
evaluateNF_ = evaluateWHNF . rnf
```

File: universum.cabal [hunk 6]
```diff
@@ -1,5 +1,5 @@
 name:                universum
-version:             0.3
+version:             0.3.1
```
Comment (neongreen) [inline universum.cabal]: 0.4?

File: src/Universum.hs (hunk 4 above)
Comment (ChShersh) [inline src/Universum.hs]: Okay, I agree that we can remove it. Also never used it.

File: src/Print.hs (hunk 3 above)
Comment (ChShersh) [inline src/Print.hs]: Already fixed in later commits.

File: universum.cabal (hunk 6 above)
Comment (ChShersh) [inline universum.cabal]: Already fixed

**Comment by jagajaga** [comment]
LGTM

File: src/Debug.hs [hunk 7]
```diff
@@ -24,43 +27,52 @@ import           System.IO.Unsafe (unsafePerformIO)
...
 {-# WARNING traceId "'traceId' remains in code" #-}
 traceId :: Text -> Text
 traceId s = trace s s
 
+-- | Similar to 'undefined' but data type.
 {-# WARNING Undefined "'Undefined' type remains in code" #-}
 data Undefined = Undefined
     deriving (P.Eq, P.Ord, P.Show, P.Read, P.Enum, P.Bounded, Data, Typeable, Generic)
 
+-- | 'P.undefined' that leaves warning in code on every usage.
```
Comment (gromakovsky) [inline src/Debug.hs]: Should `P` be here? Will haddock reader understand what this `P` means?

File: src/Universum.hs [hunk 8]
```diff
@@ -1,13 +1,21 @@
...
+-- @
+--     {-# LANGUAGE NoImplicitPrelude #-}
+--
+--     import Universum
+-- @
+
 module Universum
        ( -- * Reexports from base and from modules in this repo
-         module X
+         module X  -- Should I expand this to all modules to remove haddock warnings?
```
Comment (gromakovsky) [inline src/Universum.hs]: I don't know.

File: README.md [hunk 9]
```diff
@@ -4,105 +4,215 @@ Universum
...
+
+Structure of this tutorial
+--------------------------
+
+This tutorial has several parts:
+
+1. Philosophy and motivation.
+2. How to use it.
+3. What we've changed in `Prelude` (some gotchas).
+4. What already known things but that are not in `Prelude` we've brought into scope.
```
Comment (gromakovsky) [inline README.md]: I don't think this sentence is correct. Perhaps, you intended to write
> What already known things that are not in `Prelude` we've brought into scope.

File: README.md [hunk 10]
```diff
@@ -4,105 +4,215 @@ Universum
...
+5. What new thing we've added.
```
Comment (gromakovsky) [inline README.md]: Thing**s** maybe?

File: README.md [hunk 11]
```diff
@@ -4,105 +4,215 @@ Universum
...
 
-Lists
------
+What's new?
+-----------
 
-We export some utility functions:
+Finally, we can move to part describing what new cool features we bring with `universum`.
 
+* `foreach` is `flip fmap`.
```
Comment (gromakovsky) [inline README.md]: Didn't we remove it? I am a bit lost in changes.

Comment (gromakovsky) [inline README.md]: Ok, I see, I was really lost, this message was removed.

File: src/Debug.hs (hunk 7 above)
Comment (neongreen) [inline src/Debug.hs]: Haddock reader won't see `P`, they'll see a link to `Prelude.undefined` without a module name.

Comment (gromakovsky) [inline src/Debug.hs]: 🆒

Comment (neongreen) [inline src/Debug.hs]: Oh, you probably meant “Haddock parser”. Yes, Haddock is smart enough to understand it.

Comment (gromakovsky) [inline src/Debug.hs]: No, I actually meant a human and I didn't know that there will be a link (actually I heard about it, but didn't think about it).

**Comment by ChShersh** [comment]
@gromakovsky Some typos are fixed. I think rest typos can be fixed incrementally and they are not really so important.

File: src/Lifted/File.hs [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+getContents = liftIO XIO.getContents
+{-# INLINE getContents #-}
+
+-- | Lifted version of 'Data.Text.getLine'.
+getLine :: MonadIO m => m Text
+getLine = liftIO XIO.getLine
+{-# INLINE getLine #-}
+
+-- | Lifted version of 'Data.Text.interact'.
+interact :: MonadIO m => (Text -> Text) -> m ()
```
Comment (Martoon-00) [inline src/Lifted/File.hs]: Maybe better to use lazy version?

Comment (neongreen) [inline src/Lifted/File.hs]: And same for `getContents`.

**Comment by ChShersh** [comment]
@neongreen Tell me whether it's ok and I'll merge this to `master` eventually.

**Comment by neongreen** [comment]
I think you can merge it already.

**Pull request merged**